pallene/parser.lua

```diff
@@ -589,6 +562,39 @@ function Parser:FuncArgs()
     end
 end
 
+function Parser:FuncParams()
+    local oparen = self:e("(")
+    local params = self:DeclList()
+    local _ = self:e(")", oparen)
+    return params
+end
+
+function Parser:FuncExp(func_token, name_exp, is_stat)
+    local params = self:FuncParams()
+    local rt_types = {}
+    if self:peek(":") then
+        self:region_begin()
+        self:e()
+        rt_types = self:RetTypes()
+        self:region_end()
+    end
+
+    local block    = self:Block()
+    local _        = self:e("end", func_token)
+
+    local arg_types = {}
+    for i, decl in ipairs(params) do
+        if not decl.type and is_stat then
+            self:syntax_error(decl.loc,
+                    "Parameter '%s' is missing a type annotation", decl.name)
+        end
+      arg_types[i] = decl.type
+    end
+
+    local func_typ = ast.Type.Function(name_exp.loc, arg_types, rt_types)
+    return ast.Exp.Lambda(name_exp.loc, params, block), func_typ
```

Comment:
Do we need to receive name_exp as a parameter? We could use the func_token to get the location

Reply:
I kept it so that the function's name is used for error reporting when parsing function statements. Should I change it?

Reply:
I think this is another reason why we should put the error logic in the FuncStat.

Reply:
Ah, okay.